slam-libraries/viam-orb-slam3/orbslam_server_v1.cc

```diff
@@ -439,6 +439,38 @@ ::grpc::Status SLAMServiceImpl::GetInternalState(
     }
 }
 
+::grpc::Status SLAMServiceImpl::GetPointCloudMapStream(
+    ServerContext *context, const GetPointCloudMapStreamRequest *request,
+    ServerWriter<GetPointCloudMapStreamResponse> *writer) {
+    return ::grpc::Status(::grpc::StatusCode::UNIMPLEMENTED, "");
+}
+
+::grpc::Status SLAMServiceImpl::GetInternalStateStream(
+    ServerContext *context, const GetInternalStateStreamRequest *request,
+    ServerWriter<GetInternalStateStreamResponse> *writer) {
+    std::stringbuf buffer;
+    // deferring reading the osa file in chunks until we run into issues
+    // with loading the file into memory
+    bool success = ArchiveSlam(buffer);
+    if (!success)
+        return grpc::Status(grpc::StatusCode::UNAVAILABLE,
+                            "SLAM is not yet initialized");
+
+    std::string internal_state_chunk;
+    GetInternalStateStreamResponse response;
+    for (int start_index = 0; start_index < buffer.str().size();
```

Comment:
We should avoid calling `buffer.str().size()` on every iteration of the loop, since I believe that will copy the buffer to a string on each iteration. Instead, we can call this once before the loop starts, and save it as a const.

Reply:
@tessavitabile 
Does that mean we should also not call `buffer.str().substr` here as well https://github.com/viamrobotics/slam/pull/149/files#diff-b68d96da1bc970406f0f4402811d582812550e7c8450a636d73540f0363c73d8R464 ?

Should we should instead do something like:

```
string buffer_str = buffer.str()
for (int start_index = 0; start_index < buffer_str.size(); start_index += maximumGRPCByteChunkSize) {
internal_state_chunk = buffer_str.substr(start_index, maximumGRPCByteChunkSize);
...
}
```

Reply:
yes

Reply:
c++ question, can I use `std::move(buffer.str())` instead of making a copy? since `buffer` would not get used in the rest of the code

Reply:
I don't really know much about `std::move` but from reading the docs: https://en.cppreference.com/w/cpp/utility/move I do think that it should be safe to change `std::string buffer_str = buffer.str();` to `std::string buffer_str = std::move(buffer.str());` given you don't use it again.

Reply:
If you do change it, please run the manual tests to confirm that it hasn't changed the behavior.

Reply:
I think since the `string buffer_str = buffer.str()` includes the conversion to string, move is not actually duplicating the memory allocation. The conversion creates a string object that is then assigned to buffer_str. `std::move()` seems to be used more when copying variables of the same type and just want to send the pointer instead of the full object.

Reply:
Yes, std::move isn't necessary here, since buffer.str() is already an rvalue (unnamed value), so it will not be copied when it is assigned to buffer_str. std::move is when you want to move ownership from one lvalue (named value) to another. This doesn't share a pointer or create shared ownership--instead the moved-from value will be unsafe to access.

Reply:
kk won't make the change then, thanks!